feat/3866: document new InfluxDB OSS --hardening-enabled feature (#3867)
influxdata/docs-v2: content/influxdb/v2.1/security/enable-hardening.md

---
title: Enable security features
seotitle: Enable security/hardening features
description: >
  Enable additional security features in InfluxDB.
weight: 102
menu:
  influxdb_2_1:
    parent: Security & authorization
influxdb/v2.1/tags: [security, hardening]
products: [oss]
---

InfluxDB is used in many environments and depending on your environment, enabling additional security features may be appropriate.
To enable all security hardening options now and in the future,
start InfluxDB with the `--hardening-enabled` command line flag:

>     ```bash

Reviewer: Just a nit -- is it necessary (or conventional) to multi-line these commands when they're so short?

>     ```bash

Author: I copied this format from the TLS section, but it is reformatted away now anyway.

> InfluxDB is used in many environments and depending on your environment, enabling additional security features may be appropriate.
> To enable all security hardening options now and in the future,
> start InfluxDB with the `--hardening-enabled` command line flag:
> 
>     ```bash
>     influxd \
>     --hardening-enabled
>     ```

Reviewer 2: ```suggestion
InfluxDB {{< current-version >}} provides optional security features that ensure your
InfluxDB instance is secure in whatever environment it's used in.

To enable all [additional security features](#security-features), use the
[`hardening-enabled` configuration option](/influxdb/v2.1/reference/config-options/#hardening-enabled)
when starting InfluxDB.
```

There are three different ways to configure `influxd`; the CLI flag is only one. In the Configuration options doc, we show all three methods for setting config options. This also brings up additional changes that need to be made:

- Add a `hardening-enabled` section to the configuration options doc (`content//influxdb/v2.1/reference/config-options/`). Because this isn't available in 2.1 yet, we need to leave it commented out until 2.1 gets patched. We can port the change to the `oss-2.2-base` branch and uncomment the section.
- Update `data/influxd_flags.yml` with the new flag:

    ```yaml
    - flag: "--flux-log-enabled"
      added: 2.0

    - flag: "--hardening-enabled"
      added: 2.2

    - flag: "-h, --help"
      added: 2.0
      nolink: true
    ```

Author: > * Add a `hardening-enabled` section to the configuration options doc (`content//influxdb/v2.1/reference/config-options/`). Because this isn't available in 2.1 yet, we need to leave it commented out until 2.1 gets patched. We can port the change to the `oss-2.2-base` branch and uncomment the section.

@sanderson - what is the recommended way to comment out things? I read https://github.com/influxdata/docs-v2/blob/master/CONTRIBUTING.md but didn't see anything (maybe I missed it?).

Reviewer 2: @jdstrand Markdown just uses HTML comments `<!--  -->`.

Oh, this reminds me... don't comment out the line item in middle of the the TOC list. It won't work. For now, just add it as a comment below the list with a note to move it into the list once 2.1 gets patched.

## Security features

- [Private IP Validation](#private-ip-validation)

### Private IP Validation

> Certain flux functions (eg, `to()`, `from()`, etc) and [template fetching](/influxdb/v2.1/influxdb-templates/) cause InfluxDB to make HTTP requests over the network.
> When private IP validation is enabled, InfluxDB will first verify that the IP address of the URL is not a private IP address.
> IP addresses are considered private if they fall into one of the following categories:
> * IPv4 loopback (`127.0.0.0/8`)
> * RFC1918 (`10.0.0.0/8`, `172.16.0.0/12`, `192.168.0.0/16`)
> * RFC3927 (`169.254.0.0/16`)
> * IPv6 loopback (`::1/128`)
> * IPv6 link-local (`fe80::/10`)
> * IPv6 unique local (`fc00::/7`)

Reviewer 2: ```suggestion
Some Flux functions ([`to()`](/flux/v0.x/stdlib/influxdata/influxdb/to/),
[`from()`](/flux/v0.x/stdlib/influxdata/influxdb/from/), [`http.post()`](/flux/v0.x/stdlib/http/post/), etc.)
and [template fetching](/influxdb/v2.1/influxdb-templates/) can require InfluxDB to make
HTTP requests over the network.
With private IP validation enabled, InfluxDB first verifies that the IP address of the URL is not a private IP address.

IP addresses are considered private if they fall into one of the following categories:

- IPv4 loopback (`127.0.0.0/8`)
- RFC1918 (`10.0.0.0/8`, `172.16.0.0/12`, `192.168.0.0/16`)
- RFC3927 (`169.254.0.0/16`)
- IPv6 loopback (`::1/128`)
- IPv6 link-local (`fe80::/10`)
- IPv6 unique local (`fc00::/7`)
```

{{% note %}}
#### Private IP considerations
If your environment requires that these authenticated HTTP requests be made to private IP addresses,
omit the use of `--hardening-enabled` and

> consider instead setting up egress firewalling to limit which hosts InfluxDB is allowed to connect.

Reviewer 2: ```suggestion
consider instead setting up egress firewalling to limit which hosts InfluxDB is allowed to connect.
{{% /note %}}
```

> consider instead setting up egress firewalling to limit which hosts InfluxDB is allowed to connect.

Reviewer: Perfect!
